### PR TITLE
refactor(core): type `ComponentRef`, `ComponentFactory` and `Componen…

### PR DESCRIPTION
**Pull request by tbosch**

### CODE REVIEW

File: npm-shrinkwrap.json [hunk 1]
```diff
@@ -8503,8 +8503,9 @@
       }
     },
     "ts2dart": {
-      "version": "0.9.9",
-      "from": "ts2dart@0.9.9",
+      "version": "0.9.10",
+      "from": "ts2dart@>=0.9.9 <0.10.0",
```
Comment (mhevery) [inline npm-shrinkwrap.json]: The fact that I see a range here seems odd to me.

File: modules/angular2/src/core/application_ref.ts [hunk 2]
```diff
@@ -94,8 +94,8 @@ export function getPlatform(): PlatformRef {
  * Shortcut for ApplicationRef.bootstrap.
  * Requires a platform the be created first.
  */
-export function coreBootstrap(injector: Injector,
-                              componentFactory: ComponentFactory): ComponentRef {
+export function coreBootstrap<C>(injector: Injector,
+                                 componentFactory: ComponentFactory<C>): ComponentRef<C> {
```
Comment (mhevery) [inline modules/angular2/src/core/application_ref.ts]: Love this change!

**Comment by mhevery** [comment]
LGTM, love it.

File: npm-shrinkwrap.json (hunk 1 above)
Comment (tbosch) [inline npm-shrinkwrap.json]: This was produced by running `tools/npm/reshrinkwrap`. I could change it manually, but seems ot be fine (as there is a `resolved` property afterwards with the exact number)

**Pull request closed without merging**